### PR TITLE
HCAP-1396: Create a seeding script for other tables

### PR DESCRIPTION
**Pull request by rhattie-ey**
Original ticket: [HCAP-1396](https://eydscanada.atlassian.net/browse/HCAP-1396)

### `feed-data` rewrite

Overhauled the entire script, rewriting it from the ground up.

Features:
* Encompasses all tables needed for testing
* Simpler
* More robust error handling
  * Specific logic to handle and report foreign key errors
* More human-friendly console output
* Fully compliant with strict TS
* Generalized enough to work for any table in the database
* Importable for use in wrapper scripts
* Has a `--skip-missing-tables` flag to allow population of incomplete databases (such as for server tests)

### Test data generation

[Jupyter notebook used for generation](https://gist.github.com/rhattie-ey/1efbec9b828f076ed603e050e284a611)

Created new test data CSVs for the following tables:

* cohort_participants
* cohorts
* participant_post_hire_status
* participants
* participants_status
* phases
* post_secondary_institutions
* return_of_service_status
* site_phase_allocation

### Misc

* Updated docs
* Removed the now unneeded `copy-sample-data` command
* Updated scripts, makefiles, and actions to use new scripts
* Deleted old test data from server/script/csv and server/test-data, including all XLSX files other than those in server/tests/mock
* Flagged code that is now deprecated due to this PR
* Updated Node.js to LTS release 18.15.0
* Updated tests to work with new data
* Updated GitHub actions to use only feed-data for seeding

## Next steps
* [Improve realism of participant statuses (HCAP-1464)](https://eydscanada.atlassian.net/browse/HCAP-1464)
* [Convert migration XLSX assets to CSV (HCAP-1465)](https://eydscanada.atlassian.net/browse/HCAP-1465)

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -141,10 +129,10 @@ jobs:
         run: make local-kc-run
 
       - name: Add Sites
-        run: cd server && POSTGRES_HOST=localhost npm run feed-sites sites.xlsx
+        run: cd server && POSTGRES_HOST=localhost npm run feed-sites
```
Comment (fw-noel) [inline .github/workflows/test.yml:132]: I think I'm confused about what we need now. Feed data should now be all encompassing right? Don't we want to just consistently feed data on both cypress and jest?

Comment (rhattie-ey) [reply .github/workflows/test.yml:132]: I would say that's ideal, yes - I was aiming not to disrupt the test behaviour too much, but you know what, let's just remove all automated use of feed-sites and feed-participants. Much simpler and more predictable that way.

File: Makefile [hunk 2]
```diff
@@ -44,21 +44,16 @@ migrate-down:
...
-
 seed-participants:
-	@echo "Seeding participants from server/scripts/xlsx"
+	@echo "Seeding participants from server/test-data"
 	@docker-compose -f docker-compose.dev.yml exec server npm run feed-participants $$SHEET
 
 seed-sites:
-	@echo "Seeding sites from server/scripts/xlsx"
+	@echo "Seeding sites from server/test-data"
 	@docker-compose -f docker-compose.dev.yml exec server npm run feed-sites $$SHEET
```
Comment (mmumby) [inline Makefile:54]: Do we still need additional commands for seeding sites/participants, or can this all be accomplished by `seed-data`?

Comment (mmumby) [reply Makefile:54]: Oh i think Noel already asked the same thing. Disregard.

File: README.md [hunk 3]
```diff
@@ -113,8 +113,8 @@ In the server directory, you can run the following scripts:
...
-`npm run feed-sites [filename].xlsx` | Parses a given xlsx file inside the `server/scripts/xlsx/` folder and feeds it to the `employer_sites` table. <br> If you've spun up the application using Docker Compose, i.e. `make local-run`, you can run the site seeding script with the following make command: <br> `make seed-sites [filename].xlsx` | No
-`npm run feed-participants [filename].xlsx` | Parses a given xlsx file inside the `server/scripts/xlsx/` folder and feeds it to the `participants` table. <br> If you've spun up the application using Docker Compose, i.e. `make local-run`, you can run the participant seeding script with the following make command: <br> `make seed-participants my_spreadsheet.xlsx` | No
+`npm run feed-sites [filename].csv` | Parses a given CSV file (or `employer_sites.csv`, by default) inside the `server/test-data/` folder and feeds it to the `employer_sites` table. <br> If you've spun up the application using Docker Compose, i.e. `make local-run`, you can run the site seeding script with the following make command: <br> `make seed-sites [filename].csv` | No
```
Comment (vanics) [inline README.md:116]: Nice!

**Comment by fw-noel** [comment]
I have one test to try and resolve, but in the meantime, couple of comments
Some tests have heavy reliance on specific ids which caused a few of the breaks. In my commit, you'll see that 26, 27, and 28 were important ones. We'll want to improve this in the future by just moving the necessary participants through in the tests themselves.

The other issues are that all the csvs start with 0 ids, they should start at 1

Also, massivejs `insert` doesn't add to the sequence of a table, so there needs to be a way to ensure that we can save in the future into tables with auto generated pkeys. I added
> dbClient.runRawQuery(`ALTER SEQUENCE ${table}_id_seq RESTART WITH ${rowCount}`);
To the feed-data script, but I don't know if it's the best way.

File: server/scripts/feed-data.ts [hunk 4]
```diff
@@ -1,139 +1,182 @@
...
+ * @returns Array of `InsertResult`s for each transaction
+ */
+function insertCSV(filePath: string, table: string) {
+  return new Promise<InsertResult[]>((resolve, reject) => {
+    const rowResults: Promise<InsertResult>[] = [];
+    parseFile(filePath, { headers: true })
+      .on('error', reject)
+      .on('data', (row) => rowResults.push(insertRow(row, table)))
+      .on('end', (rowCount: number) => {
+        dbClient.runRawQuery(`ALTER SEQUENCE ${table}_id_seq RESTART WITH ${rowCount}`);
```
Comment (rhattie-ey) [inline server/scripts/feed-data.ts:101]: Add a comment explaining why this is needed - it's not a bad change, but not immediately clear.

File: server/tests/util/mock.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Function to save multiple sites worth of mock data.
+ * NOTE: this should probably be replaced with existing logic from in-use services as some point.
+ * @param sitesArg Site or array of sites to save
+ * @returns Array of responses
+ */
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const saveSites = async (sitesArg) => {
```
Comment (mmumby) [inline server/tests/util/mock.ts:12]: How much work would it be to use `saveSingleSite` to create the mocks and remove this function?

File: server/tests/participant.service.test.js [hunk 6]
```diff
@@ -167,61 +155,13 @@ describe('Participants Service', () => {
...
-      { id: 6488696, status: 'Duplicate' },
-      { id: 6488697, status: 'Duplicate' },
-      { id: 6488698, status: 'Duplicate' },
-      { id: 6488699, status: 'Duplicate' },
-    ];
-
-    expect(res).toEqual(expectedRes);
+  beforeAll(async () => {
+    await startDB();
+    await Promise.all(allParticipants.map(async (participant) => makeParticipant(participant)));
```
Comment (mmumby) [inline server/tests/participant.service.test.js:160]: This is much nicer!

**Review by mmumby — COMMENTED** [review]
Nice work, Raelynn. I don't have any major concerns, It's a good refactor! 
The only suggestion I have would be removing the mock function to create sites, and instead, use the existing service `saveSingleSite`, A few tests are already creating mocks that way. Let me know if you think this is OOS or a quick win.

**Review by fw-noel — APPROVED** [review]

File: server/tests/util/mock.ts (hunk 5 above)
Comment (rhattie-ey) [reply server/tests/util/mock.ts]: Refactored in latest commit!

**Review by fw-noel — APPROVED** [review]

**Pull request merged**